documentation/src/test/java/snippets/PaginationTest.java

```diff
@@ -0,0 +1,58 @@
+package snippets;
+
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.Uni;
+import io.smallrye.mutiny.subscription.Cancellable;
+import io.smallrye.mutiny.test.MultiAssertSubscriber;
+import org.junit.Test;
+
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.awaitility.Awaitility.await;
+
+public class PaginationTest {
+
+    private ExecutorService executor = Executors.newSingleThreadExecutor();
+
+    @Test
+    public void test() {
+        // tag::code[]
+        PaginatedApi api = new PaginatedApi();
+
+        Multi<String> stream = Uni.createFrom()
+                .completionStage(() -> new AtomicInteger(), state -> api.getPage(state.getAndIncrement()))
+                .repeat().until(list -> ! list.isEmpty())
```

Comment:
To clarify that I'm grokking this properly.

What this does is: Call api.getPage() repeatedly. When list is not empty, call onItem(), when all finished call flatten()

Is that correct?

Reply:
The `flatten` is called on every item from the stream. 
So if your API returns:
Page 0: a, b, c
Page 1: d, e
Page 2: [] (empty)

The first call will get [a, b, c] as a single item.
Flatten takes this list and passes the contained items downstream: a, b, c.

flatten does not wait until the upstream completes, it just gets each list as they arrive.

Then the iteration reaches page 2 (empty), it stops repeating. However, the item (the empty list) is still passed downstream.

Reply:
What happens if there isn't an "empty" page at the end, or we require that?

Reply:
No; it's not required, you just adapt the `until` predicate to whatever works for the API (`null`, empty, repeat indefinitely until a failure occurs (and recover)...).

If your API returns some "links" like `rel:next`, you can check for this.
You can also limit the number of pages using `takeFirstItems(x)`